### DISCUSSION TITLE
DHT22 sensor reading issues.

### DISCUSSION THREAD

**Discussion by ADenislebel** [micropython/micropython#11344] [RP2040 / Pico]
I can read 2 DHT sensors using a PIO based low level driver I forked from danjperron/PicoDHT22. When calling the driver from a single .py named RdTpUms.py (see code below) I can read both sensors in a forever loop. My issues occur when I try calling the same function from another main.py. When I do so the first reading works but all subsequent fail. I therefore believe its interaction related to using the PIO block that is being used for the low level driver I cloned from Github. Both soft boot or restart from Thonny do not resolve the issue. Only a power down and re-power, get the code to work once only again.

I am obvious very green at this new environment but for the sake of keeping my babies warm in my cold frame I will have to bow down and re-connect the hardware to my old MicroChip Pic16 C based controller for the coming growing season. 
One last note the DHT's are being operated via a FET based level converter. I also have a logic analyzer connected to one of the sensors and sure enough once the failure occurs there is no further activity on any sensor line.

Unable to attach file as I always get a file not supported error
below is a raw cut and paste from Thonny! 

/// start of Rd_TpUms.txt
from machine import Pin
import PicoDht
from PicoDht import DHT22
import utime 
Indr = Pin(2,Pin.IN,Pin.PULL_UP)	# set io pin for inside frame
Otdr = Pin(3,Pin.IN,Pin.PULL_UP)	# set io pin for outdoor sensor
Dht_Frm = None 					# one global DHT22 id name
Dht_Out = None

# initialise io pin to sensor location
def init_snsrs():
    global Dht_Frm,Dht_Out		# re-use same global
    global Indr,Otdr
    
    Dht_Frm = DHT22(Indr,None,dht11=False)	# setup frame sensor
    Dht_Out = DHT22(Otdr,None,dht11=False)	# setup outdoor sensor
        
# function read sensor @ location 
def read_snsrs(loctn):				# loctn 0->frame 1->outdor
    global Dht_Frm,Dht_Out			# re-use same global
    
    if loctn == 0:
        temp,umid = Dht_Frm.read()		# read & load tmp/umd tuples
    if loctn == 1:
        temp,umid = Dht_Out.read()
    if temp is None:				# no value is error
        temp=88.8					# return dflt error value
        umid=99.9					# return dflt error value
    return temp,umid

# test code                    works forever if py is run on its own
if __name__ == "__main__":
    cnt = 0
    init_snsrs()				# setup sensor driver
    #umid = None
    while True:
        print('Cnt',cnt,'\r')
        for i in range(2):    
            temp,umid=read_snsrs(i)
            utime.sleep_ms(500)
            print(i,temp,umid)
        cnt+=1
        utime.sleep_ms(500)                   
/// end of Rd_TpUms.txt

# main.py code from which the issues arise
print("ColdFrame running...")
Rd_TpUms.init_snsrs()                             # initialize both frame & outdoor sensors
gc.enable()  	                                       # get the garbage collector happening

while(True):
    utime.sleep_ms(500)				# pause 1/2 sec
    if Alt_flg is True:					# alternate
        Upd_tmp_umd(0)				# frame temp/umid
        Alt_flg = False
    else:
        Upd_tmp_umd(1)				# outdoor temp/umid
        Alt_flg = True
    utime.sleep_ms(500)
    utime.sleep(2)

Thanks for caring...Denis

**Comment by ADenislebel** [comment]
I now believe my problem relates to namespace and variable resolving by the interpreter. I still don't have a fix but I think I need to re-visit the basics of Python. Coming from a C environment I just can seem to get my head wrapped around how from one py to another it the globals just don't behave like I expect.

**Reply by glenn20** [reply]
To display code in a way we can understand, follow the advice in the link at the top of the Discussions page: "[Welcome (Please read before posting)](https://github.com/orgs/micropython/discussions/9111)".

From what I can see, you should rename `Rd_TpUms.txt` to `Rd_TpUms.py` and add `import Rd_TpUms` at the top of your `main.py` file. BTW. it is a "convention" to use all lowercase for python module names - but not compulsory. Also - I can't see a definition of the `Upd_tmp_umd()` and `Upd_tmp_umd(1)` functions.

And good luck with picking up these python practicalities - it's worth it :).

**Reply by ADenislebel** [reply]
Rd_TpUms.txt was created just to get it into the post as with its .py extension I could not get it there. Its also already imported in my main.py. For now I have a work around as I chose put the Rd_TpUms stuff directly in main where it works. Once I resolve the problem I revert back to keep main.py small.
Thanks...Denis

**Reply by ADenislebel** [reply]
Yet another piece of the puzzle. After the reading fails both DHT io lines become stuck low and do go back high doing the reset function that initially was used to setup the sensor reading code. Once that occurs I have been unable to find anyway to reset the pins to input other than a complete power down of the Pico! I'm pretty much completely baffled as it's got the looks of a hardware lockup!

**Reply by ADenislebel** [reply]
Sorry a typo error... stuck low and do not go back high...

**Reply by ADenislebel** [reply]
Well I did not find why the lockup occurs. However when I shrunk the main.py to its most basic to attempt to nail the problem it was gone. Obviously I've got a long way to go in understanding the upython environment. For now I will mark this discussion as answered as I do not want anyone spending time on attempting to fix stupid. Regards Denis
